### PR TITLE
voting-body: draft

### PR DESCRIPTION
**Pull request by robert-zaremba**

### CODE REVIEW

File: voting_body/src/proposal.rs [hunk 1]
```diff
@@ -1,168 +1,154 @@
...
+use crate::VoteError;
 
-/// Simple vote: user uses his all power to vote for a single option.
-#[derive(Serialize, Deserialize, BorshDeserialize, BorshSerialize, PartialEq)]
-#[serde(crate = "near_sdk::serde")]
-pub enum Vote {
-    Abstain,
-    Yes,
-    No,
+pub enum Consent {
```
[the remark this answers is not in the document]
Comment (robert-zaremba) [reply voting_body/src/proposal.rs:12]: I will add more docs later.

File: voting_body/src/view.rs [hunk 2]
```diff
@@ -1,20 +1,89 @@
...
+     **********/
+
+    /// Returns all proposals
+    /// Get proposals in paginated view.
+    pub fn get_proposals(&self, from_index: u32, limit: u32) -> Vec<ProposalOutput> {
+        (from_index..=min(self.prop_counter, from_index + limit))
+            .filter_map(|id| {
+                self.proposals
+                    .get(&id)
+                    .map(|proposal| ProposalOutput { id, proposal })
```
[the remark this answers is not in the document]
Comment (robert-zaremba) [reply voting_body/src/view.rs:40]: I'm creating a separate task for this.

File: voting_body/src/view.rs [hunk 3]
```diff
@@ -1,20 +1,89 @@
...
+            })
+            .collect()
+    }
+
```
[the remark this answers is not in the document]
Comment (robert-zaremba) [reply voting_body/src/view.rs:44]: going to handle it in a separate task.

File: voting_body/src/view.rs [hunk 4]
```diff
@@ -1,20 +1,89 @@
...
+    pub fn number_of_proposals(&self) -> u32 {
+        self.prop_counter
+    }
+
+    pub fn is_dissolved(&self) -> bool {
+        self.dissolved
+    }
+
+    /// Returns all members with permissions
+    pub fn get_members(&self) -> MembersOutput {
```
Comment (sczembor) [inline voting_body/src/view.rs:61]: we dont have members in the contract

File: voting_body/src/view.rs [hunk 5]
```diff
@@ -1,20 +1,89 @@
...
+        let (members, permissions) = self.members.get().unwrap();
+        MembersOutput {
+            members,
+            permissions,
+        }
+    }
+
+    /// Returns permissions of a given member.
+    /// Returns empty vector (`[]`) if not a member.
+    pub fn member_permissions(&self, member: AccountId) -> Vec<PropPerm> {
```
Comment (sczembor) [inline voting_body/src/view.rs:71]: same for this, we dont have in in contract

File: voting_body/src/lib.rs [hunk 6]
```diff
@@ -1,133 +1,583 @@
...
+    pub prop_counter: u32,
+    pub proposals: LookupMap<u32, Proposal>,
+
+    /// Near amount required to create a proposal. Will be slashed if the proposal is marked as
+    /// spam.
+    pub bond: Balance,
+    /// minimum amount of members to approve the proposal
+    pub threshold: u64,
+
+    /// all times below are in miliseconds
```
Comment (sczembor) [inline voting_body/src/lib.rs:39]: should we also add a start time?

File: voting_body/src/events.rs [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+pub(crate) fn emit_vote_execute(prop_id: u32, err: ExecError) {
+    emit_event(EventPayload {
+        event: "vote-execute",
+        data: json!({ "prop_id": prop_id, "status": "failed", "reason": err }),
+    });
+}
+
+pub(crate) fn emit_executed(prop_id: u32) {
+    emit_event(EventPayload {
+        event: "execute",
```
Comment (sczembor) [inline voting_body/src/events.rs:40]:
```suggestion
        event: "executed",
```

File: voting_body/src/lib.rs [hunk 8]
```diff
@@ -1,133 +1,583 @@
...
-    /// TODO: end proposal
+    /*
+     * Queries are in view.rs
+     */
+
+    /**********
+     * TRANSACTIONS
+     **********/
+
+    /// Creates a new proposal. `start` and `end` is Unix Time in milliseconds.
```
Comment (sczembor) [inline voting_body/src/lib.rs:80]:
```suggestion
    /// Creates a new proposal. 
```

File: voting_body/src/lib.rs [hunk 9]
```diff
@@ -1,133 +1,583 @@
...
-     * PRIVATE
+     * INTERNAL
      ****************/
 
+    fn assert_proposal(&self, id: u32) -> Proposal {
+        self.proposals.get(&id).expect("proposal does not exist")
+    }
+
+    fn assert_active(&self) {
+        near_sdk::require!(!self.dissolved, "dao is dissolved");
```
Comment (sczembor) [inline voting_body/src/lib.rs:232]: we dont have self.dissolved in the contract

File: voting_body/src/view.rs [hunk 10]
```diff
@@ -1,20 +1,89 @@
...
+        self.proposals
+            .get(&id)
+            .map(|proposal| ProposalOutput { id, proposal })
+    }
+
+    pub fn number_of_proposals(&self) -> u32 {
+        self.prop_counter
+    }
+
+    pub fn is_dissolved(&self) -> bool {
```
Comment (sczembor) [inline voting_body/src/view.rs:56]: we dont have this field in the contract

**Review by sczembor — COMMENTED** [review]
there are missing fields in the contract, that are being used in the functions

File: voting_body/src/events.rs (hunk 7 above)
Comment (robert-zaremba) [reply voting_body/src/events.rs:40]: I think the event name should be in the present simple

**Review by sczembor — APPROVED** [review]
lgtm

File: voting_body/src/lib.rs (hunk 6 above)
Comment (robert-zaremba) [reply voting_body/src/lib.rs:40]: no, voting body doesn't need start time

File: voting_body/src/lib.rs [hunk 11]
```diff
@@ -1,133 +1,583 @@
...
+    ) -> Result<u32, CreatePropError> {
+        self.assert_active();
+        let storage_start = env::storage_usage();
+        let user = env::predecessor_account_id();
+
+        let now = env::block_timestamp_ms();
+        let mut new_budget = 0;
+
         self.prop_counter += 1;
+        emit_prop_created(self.prop_counter, &kind);
```
[the remark this answers is not in the document]
Comment (robert-zaremba) [reply voting_body/src/lib.rs:98]: it's not needed. If we run out of storage, the tx will fail, so no event will be emitted

File: voting_body/src/lib.rs (hunk 9 above)
Comment (robert-zaremba) [reply voting_body/src/lib.rs]: it's removed

File: voting_body/src/lib.rs [hunk 12]
```diff
@@ -1,133 +1,412 @@
...
+        assert_eq!(prop.proposal.status, ProposalStatus::Approved);
+
+        ctx.predecessor_account_id = acc(4);
+        testing_env!(ctx.clone());
+        match ctr.vote(id, Vote::Approve) {
+            Err(VoteError::NotInProgress) => (),
+            x => panic!("expected NotInProgress, got: {:?}", x),
+        }
+
+        //
```
[the remark this answers is not in the document]
Comment (robert-zaremba) [reply voting_body/src/lib.rs:322]: it's intentional to give more visual separation

**Pull request merged**